### ISSUE TITLE
Cannot Attach again after disconnect

### ISSUE THREAD

**Issue by JohanMabille** [microsoft/debugpy#460]
## Environment data

- debugpy version: 1.1.0
- OS and version: Linux 5.4.60-gentoo-x86_64
- Python version (& distribution if applicable, e.g. Anaconda): Python 3.8.6 from Miniconda
- Using VS Code or Visual Studio: No

## Actual behavior

A "disconnect" request does not seem to really terminate the debug adapter as stated [here](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_Disconnect). It looks like the adapter is still running in the background and that you can reconnect with "initialize" request.

## Expected  behavior

After a "disconnect", the debug adapter should be restarted the same way it was started the first time.

## Steps to reproduce:
1. Execute the following python code in the debugee
```python
import debugpy
debugpy.listen(5416);
```
2. Send "initiliaze" and "attach" requests.
3. After receiving an attach response, send a "disconnect request";
4. Repeat step 1. `debugpy.listen` never returns.

If I skip step 4 and repeat step 2, everything works fine.

**Comment by int19h** [comment]
`listen()` is not supposed to be called multiple times. It starts a debug server in your process, and it's a one-off operation. The server should handle new connections (only one at a time) for as long as the process is running, though.

That said, "disconnect" request indeed doesn't terminate the adapter, and that's by design. The adapter is necessary for the debug server; and the server, once started, cannot be fully excised from the process for various reasons. The intent of the "disconnect" request, in any case, is to signal to the adapter that debugging should be stopped. It handles that accordingly, by disabling tracing (breakpoints etc) as much as Python implementation details allow.

**Comment by JohanMabille** [comment]
Thanks for the quick answer!

I was surprised by this behavior because of the specification of the DAP and its difference with PTVSD, but this seems a reasonable behavior anyway. Do you think it could be worth adding a comment about this difference in this [wiki page](https://github.com/microsoft/debugpy/wiki/Switching-over-from-%60ptvsd%60-to-%60debugpy%60)? I would be glad to do it myself, but I canont open a PR on a waiki page.

**Comment by int19h** [comment]
It should be https://github.com/microsoft/debugpy/wiki/API-Reference - I've added a clarification on `listen()` there.

However, I think there's a separate issue here. You mentioned that on step 4, the second call to `listen()` never returns at all. Normally, I would expect it to complain that the port is already open and cannot be re-used. That said, it's possible to call `listen()` with another port, too, and we don't try to check for that right now. A check with a clear error message here would improve this scenario a lot.

**Comment by fabioz** [comment]
Hanging in a 2nd `listen()` call does actually seem like a bug to me.

**Comment by int19h** [comment]
Yes, that part is definitely a bug, hence why this remains open. But I'd like to know why it didn't fail because the port is already opened - it does that when I try to repro. And if I specify a different port, then it apparently succeeds (but should fail loudly).

**Comment by heartacker** [comment]
https://github.com/microsoft/debugpy/issues/870#issuecomment-1143031575

maybe this is a good solution for this situation

**Comment by heartacker** [comment]
anyupdates? we need this indeed
